File: backend-spring/src/main/java/com/nearkart/entity/FarmerEntry.java
package com.nearkart.entity;

import jakarta.persistence.*;
import lombok.*;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.UUID;

@Entity
@Table(name = "farmer_entries")
@Getter @Setter
@NoArgsConstructor @AllArgsConstructor
@Builder
public class FarmerEntry {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "user_id", nullable = false)
    private UUID userId;

    @Column(name = "company_id")
    private UUID companyId;

    @Column(name = "invoice_no", nullable = false)
    private String invoiceNo;

    @Column(name = "entry_date")
    private LocalDate entryDate;

    @Column(name = "farmer_id", nullable = false)
    private UUID farmerId;

    @Column(nullable = false)
    private String village;

    @Column(name = "kharidar_id")
    private UUID kharidarId;

    @Column(name = "product_id", nullable = false)
    private UUID productId;

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal weight;

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal rate;

    @Column(nullable = false, precision = 14, scale = 2)
    private BigDecimal amount;

    @Column(precision = 14, scale = 2)
    @Builder.Default
    private BigDecimal hamali = BigDecimal.ZERO;

    @Column(precision = 14, scale = 2)
    @Builder.Default
    private BigDecimal tawali = BigDecimal.ZERO;

    @Column(precision = 14, scale = 2)
    @Builder.Default
    private BigDecimal warai = BigDecimal.ZERO;

    @Column(name = "auto_charge", precision = 14, scale = 2)
    @Builder.Default
    private BigDecimal autoCharge = BigDecimal.ZERO;

    @Column(precision = 14, scale = 2)
    @Builder.Default
    private BigDecimal kharcha = BigDecimal.ZERO;

    @Column(name = "mobile_no")
    private String mobileNo;

    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "farmer_id", insertable = false, updatable = false)
    private Farmer farmer;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "product_id", insertable = false, updatable = false)
    private Product product;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "kharidar_id", insertable = false, updatable = false)
    private Kharidar kharidar;

    @PrePersist
    protected void onCreate() {
        OffsetDateTime now = OffsetDateTime.now();
        createdAt = now; updatedAt = now;
        if (entryDate == null) entryDate = LocalDate.now();
    }

    @PreUpdate
    protected void onUpdate() { updatedAt = OffsetDateTime.now(); }
}

File: backend/app/schemas/farmer_entry.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class FarmerEntryCreate(BaseModel):
    company_id: UUID | None = None
    invoice_no: str | None = None
    entry_date: date
    farmer_id: UUID
    village: str | None = None
    kharidar_id: UUID | None = None
    product_id: UUID
    weight: Decimal
    rate: Decimal
    amount: Decimal
    hamali: Decimal = Decimal("0")
    tawali: Decimal = Decimal("0")
    warai: Decimal = Decimal("0")
    auto_charge: Decimal = Decimal("0")
    kharcha: Decimal = Decimal("0")
    mobile_no: str | None = None


class FarmerEntryUpdate(BaseModel):
    company_id: UUID | None = None
    invoice_no: str | None = None
    entry_date: date | None = None
    farmer_id: UUID | None = None
    village: str | None = None
    kharidar_id: UUID | None = None
    product_id: UUID | None = None
    weight: Decimal | None = None
    rate: Decimal | None = None
    amount: Decimal | None = None
    hamali: Decimal | None = None
    tawali: Decimal | None = None
    warai: Decimal | None = None
    auto_charge: Decimal | None = None
    kharcha: Decimal | None = None
    mobile_no: str | None = None


class FarmerEntryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    company_id: UUID | None
    invoice_no: str | None
    entry_date: date
    farmer_id: UUID
    village: str | None
    kharidar_id: UUID | None
    product_id: UUID
    weight: Decimal
    rate: Decimal
    amount: Decimal
    hamali: Decimal
    tawali: Decimal
    warai: Decimal
    auto_charge: Decimal
    kharcha: Decimal
    mobile_no: str | None
    created_at: datetime
    updated_at: datetime
